R/module_nested_tabs.R

```diff
@@ -179,42 +181,61 @@ srv_nested_tabs.teal_module <- function(id, datasets, modules, reporter) {
       "module { deparse1(modules$label) }."
     )
   )
-  modules$server_args <- teal.transform::resolve_delayed(modules$server_args, datasets)
+  moduleServer(id = id, module = function(input, output, session) {
+    modules$server_args <- teal.transform::resolve_delayed(modules$server_args, datasets)
 
-  args <- c(list(id = id), modules$server_args)
-  if (is_arg_used(modules$server, "reporter")) {
-    args <- c(args, list(reporter = reporter))
-  }
+    args <- c(list(id = "module"), modules$server_args)
+    if (is_arg_used(modules$server, "reporter")) {
+      args <- c(args, list(reporter = reporter))
+    }
 
-  if (is_arg_used(modules$server, "datasets")) {
-    args <- c(args, datasets = datasets)
-  }
+    if (is_arg_used(modules$server, "datasets")) {
+      args <- c(args, datasets = datasets)
+    }
 
-  if (is_arg_used(modules$server, "data")) {
-    data <- .datasets_to_data(modules, datasets)
-    args <- c(args, data = list(data))
-  }
+    datanames <- if (identical("all", modules$filter) || is.null(modules$filter)) {
+      datasets$datanames()
+    } else {
+      datasets$get_filterable_datanames(modules$filter) # get_filterable_datanames adds parents if present
+    }
 
-  if (is_arg_used(modules$server, "filter_panel_api")) {
-    filter_panel_api <- teal.slice::FilterPanelAPI$new(datasets)
-    args <- c(args, filter_panel_api = filter_panel_api)
-  }
+    # trigger the data when the tab is selected
+    trigger_data <- reactiveVal(1L)
+    output$data_reactive <- renderUI({
+      lapply(datanames, function(x) {
+        datasets$get_data(x, filtered = TRUE)
+      })
```

Comment:
I will measure timing of this and in the line `268: eventReactive(trigger_data(), datasets$get_data(x, filtered = TRUE))`

I hope double evaluation of `get_data()` doesn't influence anything (filter operation suppose to be cached)

Reply:
I confirm that reactive in `FilteredDataset$dataset_filtered` is evaluated once and cached. This means that each `datasets$get_data()` takes just fractions of the seconds. So having multiple `datasets$get_data()` here doesn't have any influence on the performance.